app/src/cose.h

```diff
@@ -131,25 +300,67 @@ namespace scitt::cose
       throw COSEDecodeError("Failed to decode protected header");
     }
 
-    if (header_items[ALG_INDEX].uDataType == QCBOR_TYPE_NONE)
+    if (header_items[PROFILE_INDEX].uDataType != QCBOR_TYPE_NONE)
     {
-      throw COSEDecodeError("Missing algorithm in protected header");
+      parsed.kid = cbor::as_string(header_items[PROFILE_INDEX].val.string);
     }
-    if (header_items[KID_INDEX].uDataType != QCBOR_TYPE_NONE)
+    if (header_items[ALG_INDEX].uDataType != QCBOR_TYPE_NONE)
     {
-      parsed.kid = cbor::as_string(header_items[KID_INDEX].val.string);
+      parsed.alg = header_items[ALG_INDEX].val.int64;
     }
     if (header_items[ISSUER_INDEX].uDataType != QCBOR_TYPE_NONE)
     {
       parsed.issuer = cbor::as_string(header_items[ISSUER_INDEX].val.string);
     }
+    if (header_items[CRIT_INDEX].uDataType != QCBOR_TYPE_NONE)
+    {
+      parsed.crit = std::vector<std::variant<int64_t, std::string>>();
+      QCBORItem critItem = header_items[CRIT_INDEX];
+      if (critItem.uDataType != QCBOR_TYPE_ARRAY)
+      {
+        throw COSEDecodeError(
+          "Value type of crit in protected header is not of type array");
+      }
+      int lenOfArray = critItem.val.uCount;
+      if (lenOfArray == 0)
+      {
+        throw COSEDecodeError(
+          "Cannot have crit array of length 0 in cose protected header.");
+      }
+      QCBORDecode_EnterArrayFromMapN(&ctx, COSE_HEADER_PARAM_CRIT);
+      for (int i = 0; i < lenOfArray; i++)
```

Comment:
This is going to behave pretty badly if someone uses a indefinite length array. In that case, `lenOfArray` is `UINT16_MAX`. It is better to call `GetNext` repeatedly, until it returns `QCBOR_ERR_NO_MORE_ITEMS`.
See https://github.com/laurencelundblade/t_cose/blob/f9859dd1525623cd9de7d23ac160acec7a859ef1/src/t_cose_parameters.c#L138-L184

The `lenOfArray == 0` check should also happen against the resulting vector.

Same applies for the x5chain code.